optuna/storages/_rdb/storage.py

```diff
@@ -14,34 +14,49 @@
 from typing import Optional
 from typing import Sequence
 from typing import Set
+from typing import TYPE_CHECKING
 import uuid
 
-import alembic.command
-import alembic.config
-import alembic.migration
-import alembic.script
 import numpy as np
-from sqlalchemy import func
-from sqlalchemy import orm
-from sqlalchemy.engine import create_engine
-from sqlalchemy.engine import Engine  # NOQA
-from sqlalchemy.exc import IntegrityError
-from sqlalchemy.exc import OperationalError
-from sqlalchemy.exc import SQLAlchemyError
-from sqlalchemy.sql import functions
 
 import optuna
 from optuna import distributions
 from optuna import version
+from optuna._imports import _LazyImport
 from optuna.storages._base import BaseStorage
 from optuna.storages._base import DEFAULT_STUDY_NAME_PREFIX
-from optuna.storages._rdb import models
 from optuna.study._study_direction import StudyDirection
 from optuna.study._study_summary import StudySummary
 from optuna.trial import FrozenTrial
 from optuna.trial import TrialState
 
 
+if TYPE_CHECKING:
+    import alembic.command as alembic_command
+    import alembic.config as alembic_config
+    import alembic.migration as alembic_migration
+    import alembic.script as alembic_script
+    import sqlalchemy
+    import sqlalchemy.engine as sqlalchemy_engine
+    import sqlalchemy.exc as sqlalchemy_exc
+    import sqlalchemy.orm as sqlalchemy_orm
+    import sqlalchemy.sql.functions as sqlalchemy_sql_functions
+
+    from optuna.storages._rdb import models as optuna_rdb_models
```

Comment:
Sorry, why was it that this conditioning was needed in the first place?

Reply:
Nice question. `models.py` also imports sqlalchemy modules/classes (https://github.com/optuna/optuna/blob/a07a36e124d6523677d718819cad61628e8621e7/optuna/storages/_rdb/models.py#L5-L19).

Instead of making them lazy, I make the module itself lazy to avoid introducing `_LazyImport` to `models.py` as well. If you feel it unreasonable, I'm fine to stop this conditioning and make sqlalchemy imports in `models.py` lazy.

Reply:
Ah, my question was rather, why do we need to condition based on `TYPE_CHECKING` and why can we not always do a lazy import. I'm probably just missing something related to mypy?

Reply:
As you expected, it's for static checker. Without additional imports, all modules are treated as `Any`. It could hurts the quality of Optuna codebase. Also, with `TYPE_CHECKING`, we are backed by IDE supports (e.g. pylance).

w/o TYPE_CHECKING | w/ TYPE_CHECKING
-- | --
![Screen Shot 2022-04-24 at 19 50 11](https://user-images.githubusercontent.com/5164000/164972884-cada55f8-e288-426b-a659-245937eeb390.JPG) | ![Screen Shot 2022-04-24 at 19 50 40](https://user-images.githubusercontent.com/5164000/164972886-1b4107f1-d9d2-49ee-bf32-65dab95587b7.JPG)